Long path names break Windows build (originally: MSVC Generator can break CMake builds)

During yesterdays training @narnaud  could only build with the Ninja generator, not the MSVC one.

We need to investigate whether we can reproduce this and how to fix it.

@narnaud can you reproduce the error and provide us with the error message please?

The error log is quite long, here is the beginning:
```
[build] CUSTOMBUILD : warning : cxx-qt-lib@0.7.1: ToolExecError: command did not execute successfully (status code exit code: 2): "C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.43.34808/bin/Hostx64/x64/cl.exe" "-nologo" "-MD" "-Z7" "-Brepro" "-std:c++17" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtQuickControls2" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtCore" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtQml" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtGui" "-I" "C:/Qt/6.8.2/msvc2022_64/include" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\cxx-qt-build\\target\\crates\\cxx-qt-lib\\include" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\cxx-qt-build\\target\\crates\\cxx-qt-lib\\include\\cxx-qt-lib-internals" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out/qt-build-utils/moc" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out/qt-build-utils/moc" "-W4" "/Zc:__cplusplus" "/permissive-" "/bigobj" "-DCXX_QT_GUI_FEATURE" "-DCXX_QT_QML_FEATURE" "-DCXX_QT_QUICKCONTROLS_FEATURE" "-FoC:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\956f9a0767486f47-qquickstyle.cxxqt.o" "-c" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\cxx-qt-gen/src/quickcontrols/qquickstyle.cxxqt.cpp"qtimezone.cxx.cpp [C:\dev\rust\training-handout\rust-cxx-qt\sol-imageviewer-cxx-qt\build\ImageFilter\_cargo-build_image_filter.vcxproj]
[build] CUSTOMBUILD : warning : cxx-qt-lib@0.7.1: ToolExecError: command did not execute successfully (status code exit code: 2): "C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.43.34808/bin/Hostx64/x64/cl.exe" "-nologo" "-MD" "-Z7" "-Brepro" "-std:c++17" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtQuickControls2" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtCore" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtQml" "-I" "C:/Qt/6.8.2/msvc2022_64/include/QtGui" "-I" "C:/Qt/6.8.2/msvc2022_64/include" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\cxx-qt-build\\target\\crates\\cxx-qt-lib\\include" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\cxx-qt-build\\target\\crates\\cxx-qt-lib\\include\\cxx-qt-lib-internals" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out/qt-build-utils/moc" "-I" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out/qt-build-utils/moc" "-W4" "/Zc:__cplusplus" "/permissive-" "/bigobj" "-DCXX_QT_GUI_FEATURE" "-DCXX_QT_QML_FEATURE" "-DCXX_QT_QUICKCONTROLS_FEATURE" "-FoC:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\956f9a0767486f47-qquickstyle.cxx.o" "-c" "C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\\x86_64-pc-windows-msvc\\debug\\build\\cxx-qt-lib-bb82d48ca66ed227\\out\\cxx-qt-gen/src/quickcontrols/qquickstyle.cxx.cpp" [C:\dev\rust\training-handout\rust-cxx-qt\sol-imageviewer-cxx-qt\build\ImageFilter\_cargo-build_image_filter.vcxproj]
[build] CUSTOMBUILD : error : failed to run custom build command for `cxx-qt-lib v0.7.1` [C:\dev\rust\training-handout\rust-cxx-qt\sol-imageviewer-cxx-qt\build\ImageFilter\_cargo-build_image_filter.vcxproj]
[build]   
[build]   Caused by:
[build]     process didn't exit successfully: `C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\debug\build\cxx-qt-lib-0cd847722440e25a\build-script-build` (exit code: 1)
[build]     --- stdout
[build]     cargo::rerun-if-env-changed=QMAKE
[build]     cargo::rerun-if-env-changed=QT_VERSION_MAJOR
[build]     cargo::rerun-if-changed=include/
[build]     cargo::rerun-if-changed=include/common.h
[build]     cargo::rerun-if-changed=include/assertion_utils.h
[build]     cargo::rerun-if-changed=include/core
[build]     cargo::rerun-if-changed=include/core/qanystringview.h
[build]     cargo::rerun-if-changed=include/core/qbytearray.h
[build]     cargo::rerun-if-changed=include/core/qcoreapplication.h
[build]     cargo::rerun-if-changed=include/core/qdate.h
[build]     cargo::rerun-if-changed=include/core/qdatetime.h
...
```

But I think the interesting part is here:
```
[build] C:/dev/rust/training-handout/rust-cxx-qt/sol-imageviewer-cxx-qt/build/x64/Debug/cargo/build\x86_64-pc-windows-msvc\debug\build\cxx-qt-lib-bb82d48ca66ed227\out\cxx-qt-gen/src/quickcontrols/qquickstyle.cxx.cpp(3): fatal error C1083: Cannot open include file: 'cxx-qt-lib-internals/src/quickcontrols/qquickstyle.cxxqt.h': No such file or directory [C:\dev\rust\training-handout\rust-cxx-qt\sol-imageviewer-cxx-qt\build\ImageFilter\_cargo-build_image_filter.vcxproj]
[build]     exit code: 0
```

--- a/crates/cxx-qt-build/src/dir.rs
+++ b/crates/cxx-qt-build/src/dir.rs
@@ -21,6 +21,11 @@ pub(crate) const INCLUDE_VERB: &str = "create symlink";
 #[cfg(not(unix))]
 pub(crate) const INCLUDE_VERB: &str = "deep copy files";
 
+pub(crate) fn gen() -> PathBuf {
+    // Use a short name due to the Windows file path limit!
+    out().join("cxxqtgen")
+}
+
 // Clean a directory by removing it and recreating it.
 pub(crate) fn clean(path: impl AsRef<Path>) -> Result<()> {
     let result = std::fs::remove_dir_all(&path);
@@ -37,14 +42,26 @@ pub(crate) fn clean(path: impl AsRef<Path>) -> Result<()> {
 
 /// The target directory, namespaced by crate
 pub(crate) fn crate_target() -> PathBuf {
-    target().join("crates").join(crate_name())
+    let path = target();
+    if is_exporting_crate() {
+        path.join("crates").join(crate_name())
+    } else {
+        // If we're not exporting, use a shortened path
+        // The paths for files in the OUT_DIR can get pretty long, especially if combined with
+        // Corrosion/CMake.
+        // This is an issue, as Windows has a maximum path length of 260 characters.
+        // The OUT_DIR is already namespaced by crate name, so we don't need to prefix again.
+        // See also: https://github.com/KDAB/cxx-qt/issues/1237
+        path
+    }
 }
 
 /// The target directory, namespaced by QML module
 pub(crate) fn module_target(module_uri: &str) -> PathBuf {
     module_export(module_uri).unwrap_or_else(|| {
         out()
-            .join("qml_modules")
+            // Use a short name due to the Windows file path limit!
+            .join("cxxqtqml")
             .join(module_name_from_uri(module_uri))
     })
 }
@@ -75,7 +92,8 @@ pub(crate) fn target() -> PathBuf {
         return export;
     }
 
-    out().join("cxx-qt-build").join("target")
+    // Use a short name due to the Windows file path limit!
+    out().join("cxxqtbuild")
 }
 
 /// The export directory, if one was specified through the environment.

--- a/crates/cxx-qt-build/src/lib.rs
+++ b/crates/cxx-qt-build/src/lib.rs
@@ -264,7 +264,7 @@ fn generate_cxxqt_cpp_files(
     header_dir: impl AsRef<Path>,
     include_prefix: &str,
 ) -> Vec<GeneratedCppFilePaths> {
-    let cxx_qt_dir = dir::out().join("cxx-qt-gen");
+    let cxx_qt_dir = dir::gen();
     std::fs::create_dir_all(&cxx_qt_dir).expect("Failed to create cxx-qt-gen directory!");
     std::fs::write(cxx_qt_dir.join("include-prefix.txt"), include_prefix).expect("");
 

--- a/crates/cxx-qt-lib-extras/build.rs
+++ b/crates/cxx-qt-lib-extras/build.rs
@@ -80,7 +80,7 @@ fn main() {
     });
     println!("cargo::rerun-if-changed=src/assertion_utils.h");
 
-    builder
-        .include_prefix("cxx-qt-lib-extras-internals")
-        .build();
+    // Use a short name due to the Windows file path limit!
+    // We don't re-export these headers anyway.
+    builder.include_prefix("private").build();
 }

--- a/crates/cxx-qt-lib/build.rs
+++ b/crates/cxx-qt-lib/build.rs
@@ -352,7 +352,9 @@ fn main() {
         .reexport_dependency("cxx-qt");
 
     let mut builder = CxxQtBuilder::library(interface)
-        .include_prefix("cxx-qt-lib-internals")
+        // Use a short name due to the Windows file path limit!
+        // We don't re-export these headers anyway
+        .include_prefix("private")
         .initializer(qt_build_utils::Initializer {
             file: Some("src/core/init.cpp".into()),
             ..qt_build_utils::Initializer::default_signature("init_cxx_qt_lib_core")

